### ISSUE TITLE
Plot failure with empty clusters

### ISSUE THREAD

**Issue by fabio-t** [ncborcherding/scRepertoire#42]
It so happens that I have a Seurat object with "empty" clusters. They weren't empty to begin with, but they are empty with regard to TCR. Essentially all the cells in those clusters are NOT clonotypes.

Some of the plots seem to be handling this well. `occupiedsRepertoire`, `clonalHomeostasis`, `clonalDiversity` are among these.

However, I'm unable to run `clonalProportion` and `clonalOverlap` on this object (also `compareClonotypes`). Removing the culprits from the output of `expression2List` resolves the problem.

**Comment by ncborcherding** [comment]
Hey fabio-t,

I think this is an error in using a seurat object instead of a list of combined contigs.

In general this issue is caused by the need to have a list for clonalProportion and clonalOverlap to work. Or rather are not designed to function with a single-cell/seurat object (only `occupiedscRepertoire()` is meant to function with a single-cell object in the list you gave). I'm actually surprised that some of the other functions do - just checked and I see the way I indexed some functions allows for the passing of a seurat object inadvertently (but not other single-cell objects) and I am not sure they will function correctly for certain filtering steps. If you look at the documentation for these functions, they have the df parameter as:

>  df The product of combineTCR(), combineBCR(),  or expression2List()

The product of the latter `expression2List()` should be what you use in order to graph `clonalHomeostasis`, `clonalDiversity`, `clonalProportion` and `clonalOverlap`. 

If I am off base here, please follow up. If that is the case, would you mind giving a couple more details - like output of `sessioninfo()` and the way you are calling the functions: `clonalHomeostasis()`, `clonalDiversity()`, `clonalProportion()` and `clonalOverlap()`.

**Comment by fabio-t** [comment]
I am indeed using the output of `expression2List()`.

```
> by_cluster <- expression2List(sce, group = "cluster")

> occupiedscRepertoire(sce, x.axis = "seurat_clusters")

> clonalHomeostasis(by_cluster, cloneCall = "gene+nt")
Warning message:
Removed 10 rows containing missing values (geom_bar). 

> clonalProportion(by_cluster, cloneCall = "gene+nt")
Error in `[.data.frame`(df[[i]], , 2) : undefined columns selected


> clonalDiversity(by_cluster, cloneCall = "gene+nt")
Warning message:
Removed 12 rows containing missing values (geom_point). 

> clonalOverlap(by_cluster, cloneCall="gene+nt", method="morisita")
Error in names(x) <- value : 
  'names' attribute [2] must be the same length as the vector [1]

> compareClonotypes(by_cluster, numbers = 10, cloneCall="aa", graph = "alluvial")
Error in `[.data.frame`(tbl, , 2) : undefined columns selected
```

As you can see, only some of them failed. Now, I remove the two "empty" clusters (ie, only made of NAs):

```
> by_cluster$'18' = NULL
> by_cluster$'20' = NULL

> clonalProportion(by_cluster, cloneCall = "gene+nt")

> clonalOverlap(by_cluster, cloneCall="gene+nt", method="morisita")
Warning message:
Removed 231 rows containing missing values (geom_text). 

> compareClonotypes(by_cluster, numbers = 10, cloneCall="aa", graph = "alluvial")
```

And those that failed, now work.

**Comment by fabio-t** [comment]
(the exception is of course `occupiedscRepertoire` but it was still worth mentioning that it didn't care about the empty clusters)

**Comment by ncborcherding** [comment]
Hey Fabio,

Yes I see what you are saying now. Yes it appears that some functions will require you to remove the null clusters to graph the data. There is no built in check to see if your grouping parameter has values across all groups. I will add it to the list for improvements. 

Thanks for pointing that out and I'm glad you found a way around the issue.

**Comment by fabio-t** [comment]
No problem, and thank you for this package - it has saved me a lot of time :)

While we are here though, I need to point out something (apologies for not opening another issue). The colour scheme for `clonalHomeostasis` is inverted compared to the rest (also in the documentation). The Hyperexpanded are dark blue and the Rare are red, while in the other plots is just generally the opposite.

I haven't found a way around it yet, I'm sure I can pull some trick to make it work, but I think you'd want to fix it in the package :)

**Comment by ncborcherding** [comment]
That was to help a mentor tell the difference between the graphs (proportion vs homeostasis) - but fair point, just fixed it and will be changed in the next commit.

Should be able to change it with addition of `scale_fill_manual` and just `rev()` the colorblind_vector().
```
colorblind_vector <- colorRampPalette(c("#FF4B20", "#FFB433", "#C6FDEC", "#7AC5FF", "#0348A6"))
clonalHomeostasis(combined, cloneCall = "gene") + 
scale_fill_manual(name = "Clonotype Group", 
                    values = rev(colorblind_vector(col)))
```

Where col is the length of your groups (default should be 5).

**Comment by fabio-t** [comment]
Indeed, that does it. Thanks!

**Comment by ncborcherding** [comment]
Hey Fabio-T,

Created a function `checkBlanks()` to prevent the issue - should go through and remove clusters/groups without clonotype information before starting the visualization. I am getting everything set up to update the bioconductor and master branch - for now, the updated version is at

```
devtools::install_github("ncborcherding/scRepertoire@refine")
```

Just FYI if you are interested in the function (based on your fix, so let me know if you want credit, I'll throw your name on the function)
```
checkBlanks <- function(df, cloneCall) {
    for (i in seq_along(df)) {
        if (length(df[[i]][,cloneCall]) == length(which(is.na(df[[i]][,cloneCall])))) {
            df[[i]] <- NULL
        } else {
            next()
        }
    }
    return(df)
}
```
Thanks for reaching out,
Nick

**Comment by fabio-t** [comment]
Hi sorry to reopen this, but it's still crashing unless I remove the empty columns myself:

```
Error in df[[i]] : subscript out of bounds
Calls: clonalHomeostasis -> checkBlanks
Execution halted
```

I've tried both with `refine` and with `master` branch.

**Comment by ncborcherding** [comment]
Hey Fabio,

No problem at all, I will look into it today. Mind giving me some more details - like a a summary of the data and what function you are calling?

I added the `checkBlanks()` internal function for your specific scenario, but apparently there is an issue with it.

Thanks,
Nick

**Comment by fabio-t** [comment]
Maybe this helps?

```
> head(by_cluster$'4')
                                    orig.ident nCount_RNA nFeature_RNA
P5_MildRecovered_AAACGGGTCCCAAGAT-1 5_GEX_SURF       3433         1188
P5_MildRecovered_AAATGCCAGGAGTTTA-1 5_GEX_SURF       5266         1604
P5_MildRecovered_AACCATGTCTAACTCT-1 5_GEX_SURF       8234         2341
P5_MildRecovered_AACTTTCCAAGCTGAG-1 5_GEX_SURF       6718         1670
P5_MildRecovered_AAGCCGCAGGGATCTG-1 5_GEX_SURF       4090         1395
P5_MildRecovered_AAGGAGCCAATCTGCA-1 5_GEX_SURF       3044         1115
                                    nCount_ADT nFeature_ADT percent.mt
P5_MildRecovered_AAACGGGTCCCAAGAT-1      10455           18   6.058841
P5_MildRecovered_AAATGCCAGGAGTTTA-1      11884           19   5.013293
P5_MildRecovered_AACCATGTCTAACTCT-1      10872           20   6.436726
P5_MildRecovered_AACTTTCCAAGCTGAG-1      12663           20   3.691575
P5_MildRecovered_AAGCCGCAGGGATCTG-1      11264           21   4.229829
P5_MildRecovered_AAGGAGCCAATCTGCA-1      10749           24   4.730618
                                    RNA.weight wsnn_res.2 seurat_clusters
P5_MildRecovered_AAACGGGTCCCAAGAT-1  0.3776257          4               4
P5_MildRecovered_AAATGCCAGGAGTTTA-1  0.4015999          4               4
P5_MildRecovered_AACCATGTCTAACTCT-1  0.9138815          4               4
P5_MildRecovered_AACTTTCCAAGCTGAG-1  0.5442061          4               4
P5_MildRecovered_AAGCCGCAGGGATCTG-1  0.4615699          4               4
P5_MildRecovered_AAGGAGCCAATCTGCA-1  0.5589432          4               4
                                    TCR_clonotype_id TCR_cdr3s_aa TCR_V_GENE
P5_MildRecovered_AAACGGGTCCCAAGAT-1             <NA>         <NA>       <NA>
P5_MildRecovered_AAATGCCAGGAGTTTA-1             <NA>         <NA>       <NA>
P5_MildRecovered_AACCATGTCTAACTCT-1             <NA>         <NA>       <NA>
P5_MildRecovered_AACTTTCCAAGCTGAG-1             <NA>         <NA>       <NA>
P5_MildRecovered_AAGCCGCAGGGATCTG-1             <NA>         <NA>       <NA>
P5_MildRecovered_AAGGAGCCAATCTGCA-1             <NA>         <NA>       <NA>
                                    TCR_D_GENE TCR_J_GENE TCR_C_GENE barcode
P5_MildRecovered_AAACGGGTCCCAAGAT-1       <NA>       <NA>       <NA>    <NA>
P5_MildRecovered_AAATGCCAGGAGTTTA-1       <NA>       <NA>       <NA>    <NA>
P5_MildRecovered_AACCATGTCTAACTCT-1       <NA>       <NA>       <NA>    <NA>
P5_MildRecovered_AACTTTCCAAGCTGAG-1       <NA>       <NA>       <NA>    <NA>
P5_MildRecovered_AAGCCGCAGGGATCTG-1       <NA>       <NA>       <NA>    <NA>
P5_MildRecovered_AAGGAGCCAATCTGCA-1       <NA>       <NA>       <NA>    <NA>
                                    CTgene CTnt CTaa CTstrict Frequency
P5_MildRecovered_AAACGGGTCCCAAGAT-1   <NA> <NA> <NA>     <NA>        NA
P5_MildRecovered_AAATGCCAGGAGTTTA-1   <NA> <NA> <NA>     <NA>        NA
P5_MildRecovered_AACCATGTCTAACTCT-1   <NA> <NA> <NA>     <NA>        NA
P5_MildRecovered_AACTTTCCAAGCTGAG-1   <NA> <NA> <NA>     <NA>        NA
P5_MildRecovered_AAGCCGCAGGGATCTG-1   <NA> <NA> <NA>     <NA>        NA
P5_MildRecovered_AAGGAGCCAATCTGCA-1   <NA> <NA> <NA>     <NA>        NA
                                    cloneType cluster
P5_MildRecovered_AAACGGGTCCCAAGAT-1      <NA>       4
P5_MildRecovered_AAATGCCAGGAGTTTA-1      <NA>       4
P5_MildRecovered_AACCATGTCTAACTCT-1      <NA>       4
P5_MildRecovered_AACTTTCCAAGCTGAG-1      <NA>       4
P5_MildRecovered_AAGCCGCAGGGATCTG-1      <NA>       4
P5_MildRecovered_AAGGAGCCAATCTGCA-1      <NA>       4
```

Just the first few lines of the "empty cluster", which is all blank.

**Comment by fabio-t** [comment]
`clonalHomeostasis(by_cluster, cloneCall = "gene+nt") + scale_fill_manual(name = "Clonotype Group", values = rev(colorblind_vector(5)))`

**Comment by ncborcherding** [comment]
Hey Fabio,

Found this issue - `checkBlanks()` was subtracting elements with all NA values while going through the loop, so by the time the loop reached the high number, there was no longer corresponding elements.

Should be fixed now under the master branch, please let me know if it works for you!

Thanks,
Nick

**Comment by fabio-t** [comment]
Seems like another error is popping up now, in plenty of places:

`Error in -nulled : invalid argument to unary operator`

I've just installed from master.

**Comment by ncborcherding** [comment]
Ok one more time - was able to replicate the error message when there are no NA values across all cells. So added an additional layer of evaluation to make sure not enter the modification process if there are no NA values in a list element. 

Should work now, but please let me know. Installing it locally I was able to overcome the error.

Thanks,
Nick

**Comment by fabio-t** [comment]
Seems to be working fine now, thanks :)

**Comment by ncborcherding** [comment]
Great thank you for following up and glad we figured it out!
